File: backend/app/models/application.py
from datetime import date, datetime
from enum import Enum

from sqlalchemy import Date, DateTime, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.core.db import Base


class ApplicationStatus(str, Enum):
    draft = "draft"
    sent = "sent"
    in_review = "in_review"
    interview = "interview"
    rejected = "rejected"
    accepted = "accepted"


class Application(Base):
    __tablename__ = "applications"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    company: Mapped[str] = mapped_column(String(255), index=True)
    position: Mapped[str] = mapped_column(String(255))
    location: Mapped[str | None] = mapped_column(String(255), nullable=True)
    status: Mapped[str] = mapped_column(String(50), default=ApplicationStatus.draft.value)
    applied_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    job_url: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

File: frontend/types/application.ts
export type ApplicationStatus =
  | "draft"
  | "sent"
  | "in_review"
  | "interview"
  | "rejected"
  | "accepted"

export interface Application {
  id: number
  company: string
  position: string
  location: string | null
  status: ApplicationStatus
  applied_date: string | null
  job_url: string | null
  notes: string | null
  created_at: string
  updated_at: string
}

export interface CreateApplicationPayload {
  company: string
  position: string
  location?: string | null
  status?: ApplicationStatus
  applied_date?: string | null
  job_url?: string | null
  notes?: string | null
}

export interface UpdateApplicationPayload {
  company?: string
  position?: string
  location?: string | null
  status?: ApplicationStatus
  applied_date?: string | null
  job_url?: string | null
  notes?: string | null
}